include/libopencm3/swm050/timer.h

```diff
@@ -0,0 +1,233 @@
+/** @defgroup timer_defines Timer Defines
+ *
+ * @brief <b>Defined Constants and Types for the SWM050 Timer</b>
+ *
+ * @ingroup SWM050_defines
+ *
+ * LGPL License Terms @ref lgpl_license
+ */
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2020 Caleb Szalacinski <contact@skiboy.net>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**@{*/
+#ifndef LIBOPENCM3_TIMER_H
+#define LIBOPENCM3_TIMER_H
+#include <libopencm3/cm3/common.h>
+#include <libopencm3/swm050/memorymap.h>
+
+/* Timer select */
+/** @defgroup timer_select Timer Select
+@{*/
+#define TIMER_SE0	TIMER_SE0_BASE
+#define TIMER_SE1	TIMER_SE1_BASE
+/*@}*/
+
+/* Timer level definitions */
+/** @defgroup timer_level Timer Level
+@{*/
+enum timer_level {
+	TIMER_LEVEL_LOW,
+	TIMER_LEVEL_HIGH
+};
+/*@}*/
+
+/* Timer edge mode definitions */
+/** @defgroup timer_edge_modes Timer Edge Modes
+@{*/
+enum timer_edge_modes {
+	/* Trigger on rising edge */
+	TIMER_EDGE_RISING,
+	/* Trigger on falling edge */
+	TIMER_EDGE_FALLING
+};
+/*@}*/
+
+/* Timer operation mode definitions */
+/** @defgroup timer_operation_modes Timer Operation Modes
+@{*/
+enum timer_operation_modes {
+	TIMER_MODE_COUNTER,
+	TIMER_MODE_PWM,
+	TIMER_MODE_PULSE_CAPTURE,
+	TIMER_MODE_DUTY_CYCLE_CAPTURE
+};
+/*@}*/
+
+/* Timer clock source definitions */
+/** @defgroup timer_clk_src Timer Clock Source
+@{*/
+enum timer_clk_src {
+	TIMER_CLK_INTERNAL,
+	TIMER_CLK_EXTERNAL
+};
+/*@}*/
+
+/* Timer interrupt mask definitions */
+/** @defgroup timer_int_masked Timer Interrupt Mask
+@{*/
+enum timer_int_masked {
+	TIMER_UNMASKED,
+	TIMER_MASKED
+};
+/*@}*/
+
+/* Timer loop mode definitions */
+/** @defgroup timer_loop_modes Timer Loop Modes
+@{*/
+enum timer_loop_modes {
+	TIMER_LOOP_MODE,
+	TIMER_SINGLE_MODE
+};
+/*@}*/
+
+/* Timer output mode definitions */
+/** @defgroup timer_output_modes Timer Output Modes
+@{*/
+enum timer_output_modes {
+	TIMER_OUTPUT_NONE,
+	TIMER_OUTPUT_INVERT,
+	TIMER_OUTPUT_HIGH,
+	TIMER_OUTPUT_LOW
+};
+/*@}*/
+
+/* Timer PWM period definitions */
+/** @defgroup timer_pwm_period Timer PWM Periods
+@{*/
+enum timer_pwm_period {
+	TIMER_PERIOD_0,
+	TIMER_PERIOD_1
+};
+/*@}*/
+
+/* Timer clock divider mask */
+/** @defgroup timer_div_mask Timer Clock Divider Mask
+@{*/
+#define TIMER_DIV_MASK			~(0x3F << 16)
```

Comment:
ok, this is something I didn't notice earlier.  We definitely don't store inverted masks.  All the masks are the natural mask, so 0x3f.  We generally also keep them unshifted, except for single bits, but I'm ok with you keeping them shifted in your own target here if you like, but the masks are for masking user variables not for internally clearing other fields.  Look at the inconsistency you create even in this file, in for instance:

```
SYSCTL_SYS_CFG_1 &= TIMER_DIV_MASK | ((~0x01 & div) << 16);
vs
TIMER_CTRL(timer) &= ~TIMER_CTRL_EN;
```

have a look at say, "ack '&=' lib/stm32/common" for how consistently this is normally done.

Reply:
I'm afraid I don't feel comfortable making this change myself locally though

Reply:
Fixed.